### ISSUE TITLE
`OMPTaskTrans` / `InliningTrans` / ? resulting in incorrect data sharing clauses

### ISSUE THREAD

**Issue by LonelyCat124** [stfc/PSyclone#2421]
When running OMPTaskTrans (with manually predone inlining trans) there's an issue with renamed variables not appearing in data sharing clauses.

When inlining momentum_u_code we get the expected behaviour:
`firstprivate(j_out_var,j_el_inner,u_e,depe,u_w,depw,v_sc,v_s,deps,v_nc,**v_n**,depn,` (there's an argument over whether these should be firstprivate or not but thats not really important for now).

When then inlining momentum_v_code, the inlining transformation renames `v_n` to `v_n_1` to prevent collision (makes sense), however the private/firstprivate clauses (and dependency clauses) on this task still use `v_n` instead of `v_n_1`. 

I think this is happening as the parent parallel clause doesn't seem to know anything about `v_n_1`, and thus it is considered to be `shared` by the tasks, so shouldn't be declared private or firstprivate, however this doesn't explain why the task's clauses contain `v_n` which is never used in the task, so shouldn't be inside the clauses unless renaming is happening later than I expect?

**Comment by arporter** [comment]
That's weird. By default, any symbols in outer symbol tables are in scope in any inner table. Therefore, inlining should account for *all* name clashes. So, in theory, it should work...if you could narrow down precisely where it is that would help.

**Comment by LonelyCat124** [comment]
Hi Andy,
I think the issue might be that the structure is two inlined functions with overlapping names:

```
do i = 1, 10
   call momentum_u(...) ! This contains v_n as a variable
end do

do i = 1, 10
   call momentum_v(...) ! This also contains v_n as a variable
end do
```

When these functions are inlined, they both find their scope to be the loop, so the symbols are added to that Loop's symbol table only in the `apply` function in inline_trans (since they look for `node.scope.symbol_table`). This then means the names are allowed to both exist at that time and only collide during lowering, when the Loop's individual symbol tables are merged into the routine.

The OpenMP clauses exist outside of those individual scopes, and since another symbol with the name they reference exists, PSyclone assumes the thing they're referencing to still be correct (I assume) and doesn't adjust them since they lie outside the scope being merged.

Does that make sense?

**Comment by arporter** [comment]
Ah! Yes, that makes sense. The OMP clauses need to get updated when the inlining happens. Perhaps using the new node-updating signaling mechanism.

**Comment by LonelyCat124** [comment]
My concern is I don't know that the OpenMP clauses can know how to tell what they need to use since both the original and modified named symbols can appear inside the directive's schedule.
Would it be bad to have the inline transformation search for a parent Routine and use that scope for inlining instead? Or is that somewhat bad practice in PSyclone?

**Comment by arporter** [comment]
Well, we try to avoid having to do that but sometimes we do resort to it yes. I can't remember how we generate the Clauses but could we just repeat it once the inlining is done (if it uses dep. analysis)?

**Comment by LonelyCat124** [comment]
The task one doesn't use dep analysis - at the moment the task one essentially says "please don't modify the code underneath me once i've been lowered". I guess one thing that might work would be to lower all the children of the task directive before computing the clauses - though I suspect this may not fix the issue with the OMPParallelDirective clauses. I'll try that quickly and see if that resolves the issue for now.

**Comment by LonelyCat124** [comment]
That doesn't resolve it because the merging of the symbol tables isn't happening in lowering but during code generation. I think it would be quite difficult to regenerate clauses at this point (partly as the task directive was designed in such a way that post-lowering we didn't think code changes would happen).
I could force the task region to merge all child `ScopingNode` symbol tables with its parent parallel regions scope at lowering? Would this be a reasonable option?

**Comment by arporter** [comment]
Can we add the task directives after the inlining has been performed? I think maybe I need to chat through what's going on to fully understand the situation.

**Comment by LonelyCat124** [comment]
Yeah that always will happen (though i'm doing it explicitly for now as there were some other problems with doing it after the ChunkedLoopTrans for NemoLite2D I think).

The steps I'm doing to hit this error (NemoLite2D) are:
1. Lower all Kerns to Calls
2. Inline all of the Calls
3. Apply ChunkedLoopTrans + OMPTaskTrans to the loops
4. Apply OMPParallelTrans

What seems to happen then in PSyclone is:
1. Lower all of the tree to language level - at this point the clauses on OpenMP directives are created / refreshed if needed
2. Codegen starts visiting nodes for codegen - when it reaches `routine_node` in the Fortran backend, there is code to do `whole_routine_scope.merge(schedule.symbol_table)` for every Schedule in the routine - at this point it renames a bunch of variables inside those scopes, but anything outside of those scopes (which OpenMP directives need to be by definition) doesn't understand the renames (and can't safely since the original name can be used inside the OpenMP directive still, so renaming isn't sufficient).

Happy to have a chat about it - let me know when works for you :)

**Comment by LonelyCat124** [comment]
> That doesn't resolve it because the merging of the symbol tables isn't happening in lowering but during code generation. I think it would be quite difficult to regenerate clauses at this point (partly as the task directive was designed in such a way that post-lowering we didn't think code changes would happen). I could force the task region to merge all child `ScopingNode` symbol tables with its parent parallel regions scope at lowering? Would this be a reasonable option?

This sort of works (I need to empty the SymbolTable after) but it is very "Fortran" to have to do this (local fields could remain local to that scope in other languages), but I don't have a good generic solution otherwise.

**Comment by LonelyCat124** [comment]
Also my "fix" here only works for code with tasking, if we inlined and then applied other OpenMP directives we'd get the same issue, so this should probably be done during lowering of any OpenMP parallel directive.

**Comment by arporter** [comment]
Let me see if I can write a test that exercises this...

**Comment by arporter** [comment]
I've pushed my simplified (now just generic PSyIR rather than GOcean) test to `2421_omp_clause_bug`.

**Comment by LonelyCat124** [comment]
I guess then we need to decide which fix we prefer. 

Either we say this is InlineTrans job and it has to search upwards for a Routine to merge the symbol table with, or its an issue unique to OpenMP CPU Parallelism, in which case we make it "Fortran-specific" for now and have the OMP Parallel directive merge all child ScopingNode symbol tables during lowering (before clauses are computed/updated).

**Comment by LonelyCat124** [comment]
Well its complicated because if we're being language-independent it shouldn't have to, as by the OpenMP standard variables declared inside an OpenMP region are considered to be private automatically (since they have no existence outside the region anyway its fine).

In Fortran that concept doesn't exist (local scopes) so I'm not sure what is more sensible to do. Its

**Comment by arporter** [comment]
Oooh, that is complicated. I think part of the problem is that the list of variables to put in the private clause is obtained using the `Node.reference_accesses` method and that flattens References to Signatures and assumes that if two Signatures have the same name then they are the same variable.

**Comment by arporter** [comment]
> 1. Lower all of the tree to language level - at this point the clauses on OpenMP directives are created / refreshed if needed
> 2. Codegen starts visiting nodes for codegen - when it reaches routine_node in the Fortran backend, there is code to do whole_routine_scope.merge(schedule.symbol_table) for every Schedule in the routine - at this point it renames a bunch of variables inside those scopes, but anything outside of those scopes (which OpenMP directives need to be by definition) doesn't understand the renames (and can't safely since the original name can be used inside the OpenMP directive still, so renaming isn't sufficient).

You were quite right about this (it just took me a long while to understand). In theory, the renaming should be sufficient because the clauses on the OMP directive should refer to Symbols (that retain their identity while being renamed). However, in `OMPParallelDirective.infer_sharing_attributes` we go from Symbols -> Signatures -> Symbols and thus lose any distinct Symbols that are in different tables but happen to have the same name. This needs to be solved by #2424.

**Comment by LonelyCat124** [comment]
I've been trying a bit to do inlining + OpenMP loop transformation (weird performance results) and I cannot get this to work at all. I attempted to have InlineTrans merge the scopes at a higher level (to avoid the dependence analysis failures), but this fails due to it believing one of the symbols is an argument (I think?) to a Call and refuses/believes its impossible to rename either of the colliding symbols.

I'm not quite sure why this then doesn't cause issues the way it works at the moment (where its inlined to a local scope and then merged to the routine scope later) for the tasking transformation, maybe something happens at some point during lowering/? that converts them into standard symbols.

**Comment by LonelyCat124** [comment]
@arporter I think this is the issue I mentioned yesterday.
